web-api/src/business/useCases/auth/changePasswordInteractor.ts

```diff
@@ -0,0 +1,205 @@
+import {
+  AuthFlowType,
+  ChallengeNameType,
+} from '@aws-sdk/client-cognito-identity-provider';
+import { ChangePasswordForm } from '@shared/business/entities/ChangePasswordForm';
+import { InvalidEntityError, NotFoundError } from '@web-api/errors/errors';
+import { MESSAGE_TYPES } from '@web-api/gateways/worker/workerRouter';
+import {
+  Practitioner,
+  RawPractitioner,
+} from '@shared/business/entities/Practitioner';
+import {
+  ROLES,
+  SERVICE_INDICATOR_TYPES,
+} from '@shared/business/entities/EntityConstants';
+import { RawUser, User } from '@shared/business/entities/User';
+import { ServerApplicationContext } from '@web-api/applicationContext';
+import { authErrorHandling } from '@web-api/business/useCases/auth/loginInteractor';
+import jwt from 'jsonwebtoken';
+
+export const changePasswordInteractor = async (
+  applicationContext: ServerApplicationContext,
+  {
+    code,
+    confirmPassword,
+    email,
+    password,
+    tempPassword,
+  }: {
+    password: string;
+    tempPassword?: string;
+    email: string;
+    confirmPassword: string;
+    code?: string;
+  },
+): Promise<{ idToken: string; accessToken: string; refreshToken: string }> => {
+  try {
+    const errors = new ChangePasswordForm({
+      confirmPassword,
+      email,
+      password,
+    }).getFormattedValidationErrors();
+
+    if (errors) {
+      throw new InvalidEntityError('Change Password Form Entity is invalid');
+    }
+
+    if (tempPassword) {
+      const initiateAuthResult = await applicationContext
+        .getCognito()
+        .initiateAuth({
+          AuthFlow: AuthFlowType.USER_PASSWORD_AUTH,
+          AuthParameters: {
+            PASSWORD: tempPassword,
+            USERNAME: email,
+          },
+          ClientId: applicationContext.environment.cognitoClientId,
+        });
+
+      if (
+        initiateAuthResult.ChallengeName !==
+        ChallengeNameType.NEW_PASSWORD_REQUIRED
+      ) {
+        throw new Error('User is not in `FORCE_CHANGE_PASSWORD` state');
```

Comment:
Just thinking about this implementation and pattern. Thinking about option enabling architecture could/should the persistence/auth layer handle these exceptions and throw non-Cognito like error messages? 

For example, 

```
if (tempPassword) { 
  const { AccessToken, IdToken, RefreshToken } = await applicationContext.getAuthProvider().confirmTempUser(email, tempPassword);
  // the above would throw the relevent errors

  // queue work if they're a pendingEmail
} else {
  const { AccessToken, IdToken, RefreshToken } = await applicationContext.getAuthProvider().conformForgotPassword(code, email, password);
  // the above would throw the relevent errors

}
```

What you have is totally okay for now, but trying to think about potential next steps.

Reply:
Good observation, we are tackling some of this as a part of DoD here: https://github.com/flexion/ef-cms/pull/10278